textpipe/doc.py

```diff
@@ -3,38 +3,55 @@
 Clean text, make it readable and obtain metadata from it.
 """
 
-import re
 import functools
+import re
 
-from bs4 import BeautifulSoup
 import cld2
 import spacy
 import textacy
 import textacy.text_utils
+from bs4 import BeautifulSoup
+
+from textpipe.utils import hash_dict
+
 
+class TextpipeMissingModelException(Exception):
+    """Raised when the requested model is missing"""
+    pass
 
-class Doc:
+
+class TextpipeLanguageNotSupportedException(Exception):
+    """Raised when the requested language is not in the list of allowed languages"""
+    pass
+
+
+class Doc:  # pylint: disable=too-many-arguments
     """
     Create a doc instance of text, obtain cleaned, readable text and
     metadata from this doc.
 
     Properties:
     raw: incoming, unedited text
-    clean: string containing the cleaned text
     language: 2-letter code for the language of the text
-    is_detected_language: a boolean indicating if the language was specified
-                            beforehand or detected
     hint_language: language you expect your text to be
+    spacy_nlps: dictionary containing loaded spacy language modules
+    allowed_languages: a tuple with language codes that are allowed in the current pipeline
     """
 
-    def __init__(self, raw, language=None, hint_language='en'):
+    def __init__(self, raw, language=None, hint_language='en', spacy_nlps=None,
+                 allowed_languages=('nl', 'en')):
         self.raw = raw
         self.is_detected_language = language is None
         self.hint_language = hint_language
         self._language = language
-
+        self.allowed_languages = allowed_languages
+        self.spacy_nlps = spacy_nlps or dict()
+        self._spacy_docs = {}
         self._text_stats = {}
 
+        # Set the spacy default language modules:
+        self._set_spacy_defaults()
```

Comment:
We would be losing lazy loading of SpaCy with this. If you only run `textpipe.doc.Doc('Test').language` you would now also be loading SpaCy. Is this really needed?

Reply:
No; made this lazy again.